runtime/src/accounts_cache.rs

```diff
@@ -0,0 +1,236 @@
+            ("num_slots", self.cache.len(), i64),
+            ("total_unique_writes_size", total_unique_writes_size, i64),
+        );
+    }
+
+    pub fn store(&self, slot: Slot, pubkey: &Pubkey, account: Account, hash: Hash) {
+        let slot_cache = self.slot_cache(slot).unwrap_or_else(||
+            // DashMap entry.or_insert() returns a RefMut, essentially a write lock,
+            // which is dropped after this block ends, minimizing time held by the lock.
+            // However, we still want to persist the reference to the `SlotStores` behind
+            // the lock, hence we clone it out, (`SlotStores` is an Arc so is cheap to clone).
+            self
+                .cache
+                .entry(slot)
+                .or_insert(Arc::new(SlotCacheInner::default()))
+                .clone());
+
+        slot_cache.insert(pubkey, account, hash);
+    }
+
+    pub fn load(&self, slot: Slot, pubkey: &Pubkey) -> Option<CachedAccount> {
+        self.slot_cache(slot)
+            .and_then(|slot_cache| slot_cache.get_cloned(pubkey))
+    }
+
+    pub fn remove_slot(&self, slot: Slot) -> Option<SlotCache> {
+        self.cache.remove(&slot).map(|(_, slot_cache)| slot_cache)
+    }
+
+    pub fn slot_cache(&self, slot: Slot) -> Option<SlotCache> {
+        self.cache.get(&slot).map(|result| result.value().clone())
+    }
+
+    pub fn add_root(&self, root: Slot) {
+        self.cached_roots.write().unwrap().insert(root);
+    }
+
+    pub fn clear_roots(&self) -> HashSet<Slot> {
+        std::mem::replace(&mut self.cached_roots.write().unwrap(), HashSet::new())
+    }
+
+    // Removes slots less than or equal to `max_root`. Only safe to call with roots, otherwise
+    // the slot removed could still be undergoing replay!
+    pub fn remove_slots_le(&self, max_root: Slot) -> Vec<(Slot, SlotCache)> {
+        let mut removed_slots = vec![];
+        self.cache.retain(|slot, slot_cache| {
+            let should_remove = *slot <= max_root;
+            if should_remove {
+                removed_slots.push((*slot, slot_cache.clone()))
+            }
+            !should_remove
+        });
+        removed_slots
+    }
+
+    pub fn find_older_frozen_slots(&self, num_to_retain: usize) -> Vec<Slot> {
+        if self.cache.len() > num_to_retain {
+            let mut slots: Vec<_> = self
+                .cache
+                .iter()
+                .filter_map(|item| {
+                    let (slot, slot_cache) = item.pair();
+                    if slot_cache.is_frozen() {
+                        Some(*slot)
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+            slots.sort_unstable();
+            slots.truncate(slots.len().saturating_sub(num_to_retain));
+            slots
+        } else {
+            vec![]
+        }
+    }
+
+    pub fn num_slots(&self) -> usize {
```

Comment:
nits: maybe this doesn't need to be `pub`?

Reply:
called by `self.accounts_cache.num_slots() <= MAX_CACHE_SLOTS` in `accounts_db.rs`